File: detectron_to_masks/clipper2/private/clipper_ver6.4.2/Delphi/cairo demo/cairo_clipper.pas
unit cairo_clipper;

(*******************************************************************************
*                                                                              *
* Author    :  Angus Johnson                                                   *
* Version   :  1.2                                                             *
* Date      :  29 September 2011                                               *
* Website   :  http://www.angusj.com                                           *
* Copyright :  Angus Johnson 2010-2011                                         *
*                                                                              *
* License:                                                                     *
* Use, modification & distribution is subject to Boost Software License Ver 1. *
* http://www.boost.org/LICENSE_1_0.txt                                         *
*                                                                              *
*******************************************************************************)

interface

uses
  SysUtils, Classes, Cairo, types, math, clipper;

//nb: Since Clipper only accepts integer coordinates, fractional values have to
//be scaled up and down when being passed to and from Clipper. This is easily
//accomplished by setting the scaling factor (10^x) in the following functions.
//When scaling, remember that on most platforms, integer is only a 32bit value.
function PointArrayToCairo(const polys: TPaths;
  cairo: Pcairo_t; scaling_factor: integer = 2): boolean;
function CairoToPointArray(cairo: Pcairo_t;
  out polys: TPaths; scaling_factor: integer = 2): boolean;

implementation

type
  PCairoPathDataArray = ^TCairoPathDataArray;
  TCairoPathDataArray =
    array [0.. MAXINT div sizeof(cairo_path_data_t) -1] of cairo_path_data_t;

function PointArrayToCairo(const polys: TPaths;
  cairo: Pcairo_t; scaling_factor: integer = 2): boolean;
var
  i,j: integer;
  scaling: double;
begin
  result := assigned(cairo);
  if not result then exit;
  if abs(scaling_factor) > 6 then
    raise Exception.Create('PointArrayToCairo: invalid scaling factor');
  scaling := power(10, scaling_factor);
  for i := 0 to high(polys) do
  begin
    cairo_new_sub_path(cairo);
    for j := 0 to high(polys[i]) do
      with polys[i][j] do cairo_line_to(cairo,X/scaling,Y/scaling);
    cairo_close_path(cairo);
  end;
end;
//------------------------------------------------------------------------------

function CairoToPointArray(cairo: Pcairo_t;
  out polys: TPaths; scaling_factor: integer = 2): boolean;
const
  buffLen1: integer = 32;
  buffLen2: integer = 128;
var
  i,currLen1, currLen2: integer;
  pdHdr: cairo_path_data_t;
  path: Pcairo_path_t;
  currPos: TIntPoint;
  scaling: double;
begin
  if abs(scaling_factor) > 6 then
    raise Exception.Create('PointArrayToCairo: invalid scaling factor');
  scaling := power(10, scaling_factor);
  result := false;
  setlength(polys, buffLen1);
  currLen1 := 1;
  currLen2 := 0;
  currPos := IntPoint(0,0);
  i := 0;
  path := cairo_copy_path_flat(cairo);
  try
    while i < path.num_data do
    begin
      pdHdr := PCairoPathDataArray(path.data)[i];
      case pdHdr.header._type of
      CAIRO_PATH_CLOSE_PATH:
        begin
          if currLen2 > 1 then //ie: ignore if < 3 points (not a polygon)
          begin
            setlength(polys[currLen1-1], currLen2);
            setlength(polys[currLen1], buffLen2);
            inc(currLen1);
          end;
          currLen2 := 0;
          currPos := IntPoint(0,0);
          result := true;
        end;
      CAIRO_PATH_MOVE_TO, CAIRO_PATH_LINE_TO:
        begin
          result := false;
          if (pdHdr.header._type = CAIRO_PATH_MOVE_TO) and
            (currLen2 > 0) then break; //ie enforce ClosePath for polygons
          if (currLen2 mod buffLen2 = 0) then
            SetLength(polys[currLen1-1], currLen2 + buffLen2);
          with PCairoPathDataArray(path.data)[i+1].point do
            currPos := IntPoint(Round(x*scaling),Round(y*scaling));
          polys[currLen1-1][currLen2] := currPos;
          inc(currLen2);
        end;
      end;
      inc(i, pdHdr.header.length);
    end;
  finally
    cairo_path_destroy(path);
  end;
  dec(currLen1); //ie enforces a ClosePath
  setlength(polys, currLen1);
end;
//------------------------------------------------------------------------------

end.

File: detectron_to_masks/clipper2/private/clipper_ver6.4.2/cpp/cpp_cairo/cairo_clipper.cpp
/*******************************************************************************
*                                                                              *
* Author    :  Angus Johnson                                                   *
* Version   :  1.1                                                             *
* Date      :  4 April 2011                                                    *
* Copyright :  Angus Johnson 2010-2011                                         *
*                                                                              *
* License:                                                                     *
* Use, modification & distribution is subject to Boost Software License Ver 1. *
* http://www.boost.org/LICENSE_1_0.txt                                         *
*                                                                              *
* Modified by Mike Owens to support coordinate transformation                  *
*******************************************************************************/

#include <stdexcept>
#include <cmath>
#include <cairo.h>
#include "clipper.hpp"
#include "cairo_clipper.hpp"

namespace ClipperLib {
  namespace cairo {

    namespace {

      inline cInt Round(double val)
      {
        if ((val < 0)) return (cInt)(val - 0.5); else return (cInt)(val + 0.5);
      }

      void transform_point(cairo_t* pen, Transform transform, cInt* x, cInt* y)
      {
        double _x = (double)*x, _y = (double)*y;
        switch (transform)
        {
          case tDeviceToUser:
            cairo_device_to_user(pen, &_x, &_y);
            break;
          case tUserToDevice:
            cairo_user_to_device(pen, &_x, &_y);
            break;
          default:
            ;
        }
        *x = Round(_x); *y = Round(_y);
      }
    }

    void cairo_to_clipper(cairo_t* cr,
                          Paths &pg,
                          int scaling_factor,
                          Transform transform)
    {
      if (scaling_factor > 8 || scaling_factor < 0)
        throw clipperCairoException("cairo_to_clipper: invalid scaling factor");
      double scaling = std::pow((double)10, scaling_factor);

      pg.clear();
      cairo_path_t *path = cairo_copy_path_flat(cr);

      int poly_count = 0;
      for (int i = 0; i < path->num_data; i += path->data[i].header.length) {
        if( path->data[i].header.type == CAIRO_PATH_CLOSE_PATH) poly_count++;
      }

      pg.resize(poly_count);
      int i = 0, pc = 0;
      while (pc < poly_count)
      {
        int vert_count = 1;
        int j = i;
        while(j < path->num_data &&
          path->data[j].header.type != CAIRO_PATH_CLOSE_PATH)
        {
          if (path->data[j].header.type == CAIRO_PATH_LINE_TO)
            vert_count++;
          j += path->data[j].header.length;
        }
        pg[pc].resize(vert_count);
        if (path->data[i].header.type != CAIRO_PATH_MOVE_TO) {
          pg.resize(pc);
          break;
        }
        pg[pc][0].X = Round(path->data[i+1].point.x *scaling);
        pg[pc][0].Y = Round(path->data[i+1].point.y *scaling);
        if (transform != tNone)
          transform_point(cr, transform, &pg[pc][0].X, &pg[pc][0].Y);

        i += path->data[i].header.length;

        j = 1;
        while (j < vert_count && i < path->num_data &&
          path->data[i].header.type == CAIRO_PATH_LINE_TO) {
          pg[pc][j].X = Round(path->data[i+1].point.x *scaling);
          pg[pc][j].Y = Round(path->data[i+1].point.y *scaling);
          if (transform != tNone)
            transform_point(cr, transform, &pg[pc][j].X, &pg[pc][j].Y);
          j++;
          i += path->data[i].header.length;
        }
        pc++;
        i += path->data[i].header.length;
      }
      cairo_path_destroy(path);
    }
    //--------------------------------------------------------------------------

    void clipper_to_cairo(const Paths &pg,
                          cairo_t* cr,
                          int scaling_factor,
                          Transform transform)
    {
      if (scaling_factor > 8 || scaling_factor < 0)
        throw clipperCairoException("clipper_to_cairo: invalid scaling factor");
      double scaling = std::pow((double)10, scaling_factor);
      for (size_t i = 0; i < pg.size(); ++i)
      {
        size_t sz = pg[i].size();
        if (sz < 3)
          continue;
        cairo_new_sub_path(cr);
        for (size_t j = 0; j < sz; ++j) {
          cInt x = pg[i][j].X, y = pg[i][j].Y;
          if (transform != tNone)
            transform_point(cr, transform, &x, &y);
          cairo_line_to(cr, (double)x / scaling, (double)y / scaling);
        }
        cairo_close_path(cr);
      }
    }
    //--------------------------------------------------------------------------

  }
}
